### PR TITLE
MBS-9446 fix concurrent access in MockDispatcher

### PR DESCRIPTION
**Pull request by dsvoronin**

### CODE REVIEW

File: subprojects/common/test-okhttp/src/main/kotlin/com/avito/test/http/MockDispatcher.kt [hunk 1]
```diff
@@ -29,17 +30,21 @@ class MockDispatcher(
 
         val requestData = RequestData(request)
 
-        capturers.find { it.requestMatcher.invoke(requestData) }?.run {
-            capture(request)
-            logger("request captured: $request")
+        synchronized(capturers) {
```
Comment (dsvoronin) [inline subprojects/common/test-okhttp/src/main/kotlin/com/avito/test/http/MockDispatcher.kt:33]: from docs: 

```
     * It is imperative that the user manually synchronize on the returned
     * list when iterating over it:
     * <pre>
     *  List list = Collections.synchronizedList(new ArrayList());
     *      ...
     *  synchronized (list) {
     *      Iterator i = list.iterator(); // Must be in synchronized block
     *      while (i.hasNext())
     *          foo(i.next());
     *  }
     * </pre>
     * Failure to follow this advice may result in non-deterministic behavior.
```

File: subprojects/common/test-okhttp/src/main/kotlin/com/avito/test/http/MockDispatcher.kt [hunk 2]
```diff
@@ -29,17 +30,21 @@ class MockDispatcher(
...
 
         /**
-         * last - чтобы иметь возможность перезаписывать моки, последний записанный должен выигрывать
+         * to be able to replace mocks (last one wins) (see MBS-5878)
          *
-         * @see "MBS-5878"
+         * synchronized because of concurrent iterator access (see MBS-7636)
          */
-        val matchedMock = mocks.findLast { it.requestMatcher.invoke(requestData) }
+        val matchedMock = synchronized(mocks) {
```
Comment (dsvoronin) [inline subprojects/common/test-okhttp/src/main/kotlin/com/avito/test/http/MockDispatcher.kt:45]: from `Collections.java` docs: 

```
     * It is imperative that the user manually synchronize on the returned
     * list when iterating over it:
     * <pre>
     *  List list = Collections.synchronizedList(new ArrayList());
     *      ...
     *  synchronized (list) {
     *      Iterator i = list.iterator(); // Must be in synchronized block
     *      while (i.hasNext())
     *          foo(i.next());
     *  }
     * </pre>
     * Failure to follow this advice may result in non-deterministic behavior.
```

File: subprojects/common/test-okhttp/src/test/kotlin/com/avito/test/http/MockDispatcherTest.kt [hunk 3]
```diff
@@ -54,6 +72,44 @@ internal class MockDispatcherTest {
 
         assertThat(response.getBody()?.readUtf8()).isEqualTo("First registered")
     }
+
+    /**
+     * Run with "Repeat: Until failure" option in IDEA run configuration
+     * see MBS-7636
+     */
+    @Disabled("Used only for manual debugging of concurrency problems on dispatcher internal collections")
```
Comment (dsvoronin) [inline subprojects/common/test-okhttp/src/test/kotlin/com/avito/test/http/MockDispatcherTest.kt:80]: Don't know how to write sane test here, that won't be flaky on CI

File: subprojects/common/test-okhttp/src/test/kotlin/com/avito/test/http/MockDispatcherTest.kt [hunk 4]
```diff
@@ -5,13 +5,31 @@ import okhttp3.Headers
...
 import org.junit.jupiter.api.Test
+import java.io.OutputStream
+import java.io.PrintStream
 import java.net.InetAddress
 import java.nio.charset.Charset
 
 internal class MockDispatcherTest {
 
-    private val dispatcher = MockDispatcher(logger = { println(it) })
+    //todo create TestLogger for it
```
Comment (dsvoronin) [inline subprojects/common/test-okhttp/src/test/kotlin/com/avito/test/http/MockDispatcherTest.kt:17]: created MBS-9580 for this

File: subprojects/common/test-okhttp/src/test/kotlin/com/avito/test/http/MockDispatcherTest.kt [hunk 5]
```diff
@@ -54,6 +72,44 @@ internal class MockDispatcherTest {
...
     }
+
+    /**
+     * Run with "Repeat: Until failure" option in IDEA run configuration
+     * see MBS-7636
+     */
+    @Disabled("Used only for manual debugging of concurrency problems on dispatcher internal collections")
+    @Test
+    fun `dispatcher - register and dispatching mocks - accessed concurrently`() {
+        var throwable: Throwable? = null
```
Comment (eugene-krivobokov) [inline subprojects/common/test-okhttp/src/test/kotlin/com/avito/test/http/MockDispatcherTest.kt:83]: Are visibility problems possible?
https://stackoverflow.com/a/42479801/1282707

Comment (dsvoronin) [reply subprojects/common/test-okhttp/src/test/kotlin/com/avito/test/http/MockDispatcherTest.kt:83]: maybe; it gets job done here, and not affecting ci; 
so i decided to not invest more here

Comment (eugene-krivobokov) [reply subprojects/common/test-okhttp/src/test/kotlin/com/avito/test/http/MockDispatcherTest.kt:83]: > it gets job done here

If it happens, it makes tests useless. They won't see an error.

up to you

**Review by eugene-krivobokov — APPROVED** [review]

File: subprojects/common/test-okhttp/src/main/kotlin/com/avito/test/http/MockDispatcher.kt [hunk 6]
```diff
@@ -5,15 +5,16 @@ import com.google.gson.JsonElement
...
     private val logger: (String) -> Unit = {}
 ) : Dispatcher() {
 
-    private val mocks = mutableListOf<Mock>()
+    private val mocks = Collections.synchronizedList(mutableListOf<Mock>())
 
-    private val capturers = mutableListOf<RequestCapturer>()
+    private val capturers = Collections.synchronizedList(mutableListOf<RequestCapturer>())
 
     fun registerMock(mock: Mock) {
```
Comment (sergioboi) [inline subprojects/common/test-okhttp/src/main/kotlin/com/avito/test/http/MockDispatcher.kt:19]: Adding `@Synchronized` to fun looks simpler

**Pull request merged**